src/search/pdbs/cegar.cc

```diff
@@ -0,0 +1,589 @@
+}
+
+bool CEGAR::can_add_variable_to_pattern(int index, int var) const {
+    int pdb_size = projection_collection[index]->get_pdb()->get_size();
+    int domain_size = task_proxy.get_variables()[var].get_domain_size();
+    if (!utils::is_product_within_limit(pdb_size, domain_size, max_pdb_size)) {
+        return false;
+    }
+    int added_size = pdb_size * domain_size - pdb_size;
+    return collection_size + added_size <= max_collection_size;
+}
+
+void CEGAR::add_variable_to_pattern(int collection_index, int var) {
+    const Projection &projection = *projection_collection[collection_index];
+
+    Pattern new_pattern(projection.get_pattern());
+    new_pattern.push_back(var);
+    sort(new_pattern.begin(), new_pattern.end());
+
+    unique_ptr<Projection> new_projection =
+        compute_projection(task, move(new_pattern), rng, wildcard_plans, verbosity);
+
+    collection_size -= projection.get_pdb()->get_size();
+    collection_size += new_projection->get_pdb()->get_size();
+
+    variable_to_projection[var] = collection_index;
+    projection_collection[collection_index] = move(new_projection);
+}
+
+void CEGAR::handle_flaw(const Flaw &flaw) {
+    int collection_index = flaw.collection_index;
+    int var = flaw.variable;
+    bool added_var = false;
+    auto it = variable_to_projection.find(var);
+    if (it != variable_to_projection.end()) {
+        // Variable is contained in another pattern of the collection.
+        int other_index = it->second;
+        assert(other_index != collection_index);
+        assert(projection_collection[other_index] != nullptr);
+        if (verbosity >= utils::Verbosity::VERBOSE) {
+            utils::g_log << "var" << var << " is already in pattern "
+                         << projection_collection[other_index]->get_pattern() << endl;
+        }
+        if (can_merge_patterns(collection_index, other_index)) {
+            if (verbosity >= utils::Verbosity::VERBOSE) {
+                utils::g_log << "merge the two patterns" << endl;
+            }
+            merge_patterns(collection_index, other_index);
+            added_var = true;
+        }
+    } else {
+        // Variable is not yet in the collection.
+        if (verbosity >= utils::Verbosity::VERBOSE) {
+            utils::g_log << "var" << var
+                         << " is not in the collection yet" << endl;
+        }
+        if (can_add_variable_to_pattern(collection_index, var)) {
+            if (verbosity >= utils::Verbosity::VERBOSE) {
+                utils::g_log << "add it to the pattern" << endl;
+            }
+            add_variable_to_pattern(collection_index, var);
+            added_var = true;
+        }
+    }
+
+    if (!added_var) {
+        if (verbosity >= utils::Verbosity::VERBOSE) {
+            utils::g_log << "could not add var/merge patterns due to size "
+                "limits. Blacklisting." << endl;
+        }
+        blacklisted_variables.insert(var);
```

Comment:
We initially also had local blacklists, but this was problematic for the guaranteed linear runtime of the algorithm: the global blacklist guarantees that we "handle" one variable in every iteration - either by including it into the collection or blacklisting it. Note that we think of the collection as a whole here and not of individual patterns, i.e., having variables in or not in the collection. This makes sense because the collection is disjoint. 

If we excluded a variable for a pattern only, we would lose that runtime guarantee. Additionally, the global blacklist is used to forbid adding variables to the entire collection up-front when using the multiple CEGAR algorithm.